packages/plugin-hardhat/src/utils/deploy-impl.ts

```diff
@@ -1,66 +1,154 @@
   const encodedArgs = ImplFactory.interface.encodeDeploy(opts.constructorArgs);
   const version = getVersion(unlinkedBytecode, ImplFactory.bytecode, encodedArgs);
   const layout = getStorageLayout(validations, version);
+  const fullOpts = withDefaults(opts);
+  return { provider, validations, unlinkedBytecode, encodedArgs, version, layout, fullOpts };
+}
 
-  if (opts.kind === undefined) {
-    opts.kind = inferProxyKind(validations, version);
-  }
+export async function deployProxyImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  proxyAddress?: string,
+): Promise<DeployedProxyImpl> {
+  const deployData = await getDeployData(hre, ImplFactory, opts);
+  await processProxyKind();
 
+  let currentImplAddress;
   if (proxyAddress !== undefined) {
-    await setProxyKind(provider, proxyAddress, opts);
+    // upgrade scenario
+    currentImplAddress = await getImplementationAddress(deployData.provider, proxyAddress);
   }
 
-  const fullOpts = withDefaults(opts);
+  return deployImpl(deployData, ImplFactory, opts, currentImplAddress);
 
-  assertUpgradeSafe(validations, version, fullOpts);
+  async function processProxyKind() {
+    if (opts.kind === undefined) {
+      if (proxyAddress !== undefined && (await isBeaconProxy(deployData.provider, proxyAddress))) {
+        opts.kind = 'beacon';
+      } else {
+        opts.kind = inferProxyKind(deployData.validations, deployData.version);
+      }
+    }
 
-  if (proxyAddress !== undefined) {
-    const manifest = await Manifest.forNetwork(provider);
-    const currentImplAddress = await getImplementationAddress(provider, proxyAddress);
-    const currentLayout = await getStorageLayoutForAddress(manifest, validations, currentImplAddress);
-    assertStorageUpgradeSafe(currentLayout, layout, fullOpts);
+    if (proxyAddress !== undefined) {
+      await setProxyKind(deployData.provider, proxyAddress, opts);
+    }
+
+    if (opts.kind === 'beacon') {
+      throw new BeaconProxyUnsupportedError();
+    }
+  }
+}
+
+export async function deployBeaconImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  beaconAddress?: string,
+): Promise<DeployedBeaconImpl> {
+  const deployData = await getDeployData(hre, ImplFactory, opts);
+
+  let currentImplAddress;
+  if (beaconAddress !== undefined) {
+    // upgrade scenario
+    currentImplAddress = await getBeaconImplementationAddress(beaconAddress);
+  }
+  return deployImpl(deployData, ImplFactory, opts, currentImplAddress);
+
+  async function getBeaconImplementationAddress(beaconAddress: string): Promise<string> {
+    await assertNotProxy(beaconAddress);
+
+    const IBeaconFactory = await getIBeaconFactory(hre, ImplFactory.signer);
+    const beaconContract = IBeaconFactory.attach(beaconAddress);
+    return await beaconContract.implementation();
+  }
+
+  async function assertNotProxy(address: string) {
```

Comment:
These helper functions will be useful once when we build the Truffle plugin. I would suggest trying to have them in core. But I think this can be done in the separate PR for Truffle.

Reply:
This function is mainly implementing custom error messages which do not seem to belong in core.  The helper methods used here are part of core.